contrib/mesos/pkg/scheduler/executorinfo/registry.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package executorinfo
+
+import (
+	"sync"
+
+	"fmt"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/mesos/mesos-go/mesosproto"
+	"k8s.io/kubernetes/contrib/mesos/pkg/node"
+	"k8s.io/kubernetes/contrib/mesos/pkg/scheduler/meta"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+type Registry interface {
+	New(nodename string, resources []*mesosproto.Resource) *mesosproto.ExecutorInfo
+	Get(nodename, id string) (*mesosproto.ExecutorInfo, error)
+}
+
+type registry struct {
+	// items stores executor ID keys and ExecutorInfo values
+	items map[string]*mesosproto.ExecutorInfo
+	// protects fields above
+	mu sync.RWMutex
+
+	lookupNode node.LookupFunc
+	prototype  *mesosproto.ExecutorInfo
+	client     *client.Client
+}
+
+func NewRegistry(lookupNode node.LookupFunc, prototype *mesosproto.ExecutorInfo, client *client.Client) Registry {
+	return &registry{
+		items:      map[string]*mesosproto.ExecutorInfo{},
+		lookupNode: lookupNode,
+		prototype:  prototype,
+	}
+}
+
+func (r *registry) New(nodename string, resources []*mesosproto.Resource) *mesosproto.ExecutorInfo {
+	e := proto.Clone(r.prototype).(*mesosproto.ExecutorInfo)
+	e.Resources = resources
+
+	// hostname needs of the executor needs to match that of the offer, otherwise
+	// the kubelet node status checker/updater is very unhappy
+	setCommandArgument(e, "--hostname-override", nodename, true)
+	id, _ := NewExecutorID(e)
+	e.ExecutorId = id
+
+	r.mu.RLock()
```

Comment:
it seems like this pattern of read-lock, check, write-lock, write could break if we ever decide to schedule pods in parallel. simpler to just lock the mutex, do the check-or-else-write, and then unlock. upgrading the lock here feels potentially fragile, and in this situation overkill.